src/daft-json/src/local.rs

```diff
@@ -0,0 +1,485 @@
+    Ok(fields.into())
+}
+
+/// Get the mean and standard deviation of length of lines in bytes
+fn get_line_stats_json(bytes: &[u8], n_lines: usize) -> Option<(f32, f32)> {
+    let mut lengths = Vec::with_capacity(n_lines);
+
+    let mut bytes_trunc;
+    let n_lines_per_iter = n_lines / 2;
+
+    let mut n_read = 0;
+
+    let bytes_len = bytes.len();
+
+    // sample from 0, and 75% in the file
+    for offset in [0, (bytes_len as f32 * 0.75) as usize] {
+        bytes_trunc = &bytes[offset..];
+        let pos = next_line_position(bytes_trunc)?;
+        if pos >= bytes_len {
+            return None;
+        }
+        bytes_trunc = &bytes_trunc[pos + 1..];
+
+        for _ in offset..(offset + n_lines_per_iter) {
+            let pos = next_line_position(bytes_trunc);
+            if let Some(pos) = pos {
+                lengths.push(pos);
+                let next_bytes = &bytes_trunc[pos..];
+                if next_bytes.is_empty() {
+                    return None;
+                }
+                bytes_trunc = next_bytes;
+                n_read += pos;
+            } else {
+                break;
+            }
+        }
+    }
+
+    let n_samples = lengths.len();
+    let mean = (n_read as f32) / (n_samples as f32);
+    let mut std = 0.0;
+    for &len in lengths.iter() {
+        std += (len as f32 - mean).pow(2.0)
+    }
+    std = (std / n_samples as f32).sqrt();
+    Some((mean, std))
+}
+
+/// Calculate the max number of chunks to split the file into
+/// It looks for the largest number divisible by `n_threads` and less than `chunk_size`
+/// It has an arbitrary limit of `n_threads * n_threads`, which seems to work well in practice.
+///
+/// Example:
+///
+/// ```text
+/// n_threads = 4
+/// chunk_size = 2048
+/// calculate_chunks_and_size(n_threads, chunk_size) = (16, 128)
+/// ```
+fn calculate_chunks_and_size(n_threads: usize, chunk_size: usize, total: usize) -> (usize, usize) {
+    let mut max_divisible_chunks = n_threads;
+
+    // The maximum number of chunks is n_threads * n_threads
+    // This was chosen based on some crudely done benchmarks. It seems to work well in practice.
+    // The idea is to have a number of chunks that is a divisible by the number threads to maximize parallelism.
+    // But we dont want to have too small chunks, as that would increase the overhead of the parallelism.
+    // This is a heuristic and could be improved.
+    let max_chunks = n_threads * n_threads;
+
+    while max_divisible_chunks <= chunk_size && max_divisible_chunks < max_chunks {
+        let md = max_divisible_chunks + n_threads;
+        if md > chunk_size || md > max_chunks {
+            break;
+        }
+        max_divisible_chunks = md;
+    }
+    let chunk_size = total / n_threads;
+    (max_divisible_chunks, chunk_size)
+}
+
+#[inline(always)]
+fn parse_raw_value<'a>(
+    raw_value: &'a RawValue,
+    scratch: &'a mut Vec<u8>,
+) -> std::result::Result<Value<'a>, super::Error> {
```

Comment:
Nit: This crate's result type is defined here: https://github.com/Eventual-Inc/Daft/blob/ea395ddd59d37aee024e8cfdbf18bdca755a673a/src/daft-json/src/lib.rs#L67